src/core/net/dnssd_server.cpp

```diff
@@ -668,34 +759,381 @@ const Srp::Server::Service *Server::GetNextSrpService(const Srp::Server::Host &
+        readOffset += sizeof(question);
+
+        switch (question.GetType())
+        {
+        case ResourceRecord::kTypePtr:
+            ExitNow(sdType = kDnsQueryBrowse);
+        case ResourceRecord::kTypeSrv:
+        case ResourceRecord::kTypeTxt:
+            ExitNow(sdType = kDnsQueryResolve);
+        }
+    }
+
+    for (uint16_t i = 0, readOffset = sizeof(Header); i < aHeader.GetQuestionCount(); i++)
+    {
+        Question question;
+
+        IgnoreError(Name::ReadName(aMessage, readOffset, aName, sizeof(aName)));
+        IgnoreError(aMessage.Read(readOffset, question));
+        readOffset += sizeof(question);
+
+        switch (question.GetType())
+        {
+        case ResourceRecord::kTypeAaaa:
+        case ResourceRecord::kTypeA:
+            ExitNow(sdType = kDnsQueryResolveHost);
+        }
+    }
+
+exit:
+    return sdType;
+}
+
+bool Server::HasQuestion(const Header &aHeader, const Message &aMessage, const char *aName, uint16_t aQuestionType)
+{
+    bool found = false;
+
+    for (uint16_t i = 0, readOffset = sizeof(Header); i < aHeader.GetQuestionCount(); i++)
+    {
+        Question question;
+        Error    error;
+
+        error = Name::CompareName(aMessage, readOffset, aName);
+        IgnoreError(aMessage.Read(readOffset, question));
+        readOffset += sizeof(question);
+
+        if (error == kErrorNone && aQuestionType == question.GetType())
+        {
+            ExitNow(found = true);
+        }
+    }
+
+exit:
+    return found;
+}
+
+void Server::HandleTimer(Timer &aTimer)
+{
+    aTimer.Get<Server>().HandleTimer();
+}
+
+void Server::HandleTimer(void)
+{
+    TimeMilli now = TimerMilli::GetNow();
+
+    for (QueryTransaction &query : mQueryTransactions)
+    {
+        TimeMilli expire;
+
+        if (!query.IsValid())
+        {
+            continue;
+        }
+
+        expire = query.GetStartTime() + kQueryTimeout;
+        if (expire <= now)
+        {
+            FinalizeQueryTransaction(query, Header::kResponseSuccess);
+        }
+    }
+
+    ResetTimer();
+}
+
+void Server::ResetTimer(void)
+{
+    TimeMilli now        = TimerMilli::GetNow();
+    TimeMilli nextExpire = now.GetDistantFuture();
+
+    for (QueryTransaction &query : mQueryTransactions)
+    {
+        TimeMilli expire;
+
+        if (!query.IsValid())
+        {
+            continue;
+        }
+
+        expire = query.GetStartTime() + kQueryTimeout;
+        if (expire <= now)
+        {
+            nextExpire = now;
+        }
+        else if (expire < nextExpire)
+        {
+            nextExpire = expire;
+        }
+    }
+
+    if (nextExpire < now.GetDistantFuture())
+    {
+        mTimer.FireAt(nextExpire);
+    }
+    else
+    {
+        mTimer.Stop();
+    }
+}
+
+void Server::FinalizeQueryTransaction(QueryTransaction &aQuery, Header::Response aResponseCode)
+{
+    char name[Name::kMaxNameSize];
+
+    OT_ASSERT(mQueryUnsubscribe != nullptr);
```

Comment:
```suggestion
    OT_ASSERT(mQueryUnsubscribe != nullptr);

    char name[Name::kMaxNameSize];
```

Reply:
I think we always put variable definition at the top.